### ISSUE TITLE
index.php?p=admin/actions/rich-variables keeps loading 

### ISSUE THREAD

**Issue by kgollert** [nystudio107/craft3-richvariables#29]
### Question
index.php?p=admin/actions/rich-variables keeps loading multiple times. It takes like forever. 
It looks like it's loading for each rich text field.
Is this normal? 

### Additional context
![image](https://user-images.githubusercontent.com/25514405/114064675-29e21500-989a-11eb-859c-d9650ce2ee05.png)

Setup:
![image](https://user-images.githubusercontent.com/25514405/114064879-57c75980-989a-11eb-97d9-0d38841a9d86.png)

**Comment by khalwat** [comment]
It's normal for it to load for each Rich Text field -- how many do you have in this section?

However, it should not take long at all, it's just retrieving the globals for the menu

**Comment by kgollert** [comment]
This is a long entry with ~30 rich texts fields. The Problem is the server not serving the GET requests "in time". So the queue of GETs even triggers the browser to show the "page not responding" alert (after a minute or so, the server responses are coming in). Might be a server/config issue. It's not our "prod" env.

**Comment by khalwat** [comment]
Yeah this sounds like some kind of a local development / environmental issue. The response should be immediate.

**Comment by kgollert** [comment]
Hi Andrew ( @khalwat ), me again. Checked the issue with our infrastructure guys. The server response is alright, timing is around 200-500ms for everything per GET request. 
The problem is the pure amount of requests. On one page I have tables where each cell uses a redactor field (because the editors want to click on 'bold' and insert a global var via dropdown 🤨 ). So I have like 200+ rich text fields...

Wouldn't it be possible to just load it once per page? The payload (the global vars) should be always the same. This would speed up pages with a lot of rich text fields immensely.

![image](https://user-images.githubusercontent.com/25514405/114689084-a602af80-9d15-11eb-92f8-a37266775a2e.png)

**Comment by khalwat** [comment]
Unfortunately, that's just not how Redactor plugins work. I'm not aware of a way to get it to do that loading just once per page, but I will say that 400ms or so for a simple JSON request seems _excessively_ long. It should be much, much quicker than this.

As a comparison, in my local dev environment I'm seeing about 50ms

I'll have a look to see if there's any way around this, perhaps I can figure something out to handle this specific case, such as storing the JSON response in local storage and checking there first for it.

You up for giving it a test if I have a version to try?

**Comment by kgollert** [comment]
Yes, Andrew, I'm ofc open to this. I would give a try.
on localhost (docker on OSX) it's even worse 😱 :

![image](https://user-images.githubusercontent.com/25514405/114719144-0d7d2700-9d37-11eb-87f5-b1172c55e93c.png)

**Comment by khalwat** [comment]
You should check out my dual-container setup for Docker: 

https://nystudio107.com/blog/an-annotated-docker-config-for-frontend-web-development#xdebug-performance

Anyway I will have a go at caching it in local storage.

**Comment by kgollert** [comment]
Thanks for the hint and for looking into this. 🤘

**Comment by khalwat** [comment]
Added a localstorage cache in: https://github.com/nystudio107/craft-richvariables/commit/3b3e391245be1fda309e61d030db86e2069efc09

You can try it now by setting your semver in your `composer.json` to look like this:
```json
    "nystudio107/craft-richvariables": "dev-develop as 1.0.29”,
```
Then do a `composer clear-cache && composer update`

**Comment by kgollert** [comment]
Works flawless on local. Genius 👏 
Will try it on our test-env. Thanks so much!

**Comment by khalwat** [comment]
Okay cool. I'm likely going to make one more change to it, so that the caches expire after a reasonable period of time. Right now it's cached forever (until your clear your browser's local storage cache or it does it itself), which probably isn't great.

I'll ping you to test the modified version once I have it right. TIA!

**Comment by khalwat** [comment]
Implemented a 60 second cache in: https://github.com/nystudio107/craft-richvariables/commit/6738074ea59034c212097aa082b10900ae171318

You can try it now by setting your semver in your `composer.json` to look like this:
```json
    "nystudio107/craft-richvariables": "dev-develop as 1.0.29”,
```
Then do a `composer clear-cache && composer update`

lmk...

**Comment by khalwat** [comment]
Released in 1.0.29 https://github.com/nystudio107/craft-richvariables/releases/tag/1.0.29
